src/Compiler/Checking/MethodCalls.fs

```diff
@@ -392,9 +398,11 @@ let AdjustCalledArgTypeForOptionals (infoReader: InfoReader) ad enforceNullableO
 
         // FSharpMethod(x = arg), optional F#-style argument, should have option type
         | CalleeSide ->
-            let calledArgTy2 = 
+            let calledArgTy2 =
                 if isOptionTy g calledArgTy then
                     destOptionTy g calledArgTy
+                elif g.langVersion.SupportsFeature LanguageFeature.SupportValueOptionsAsOptionalParameters && isValueOptionTy g calledArgTy then
```

Comment:
I think I saw at least 5 variants on "langVersion && isValueOptionTy" around the codebase.
Also similarities around  if-else doing mkValueOptionty in one case, mkOptionTy in another.

=> It would be good to unify the logic in one place.

Reply:
As in unify the check? Because theyre a bit different. I would personally prefer to have codegen explicit (as in keep `mk*` in place), but I can put checks in a functions.

Reply:
What about API for picking between struct/reference optional container, doing something like this:


pickOptionalContainer g calledArgTy mkOptionTy mkValueOptionTy 



let pickOptionalContainer  g innerTy refOptionMaker structOptionMaker = 
                    if g.langVersion.SupportsFeature LanguageFeature.SupportValueOptionsAsOptionalParameters && isValueOptionTy g innerTy then
                        structOptionMakerg innerTy , TypeDirectedConversionUsed.No, None
                    else
                        refOptionMaker g innerTy , TypeDirectedConversionUsed.No, None

Reply:
And the two functions passed would be altering, depending on the place:
mkValueOptionTy  + mkOptionTy 
destValueOptionTy  + destOptionTy 
mkValueNone  + mkNone 


(Not sure what other parametrization would be needed, but this seems to be structure I can see in the code blocks here)

Reply:
I can factor it out, though don't think it's worth it

Reply:
The main risk I see is forgetting one of the places in a future modification or bugfix.
If the logic is centralized, this gets more explicit.

I think in general we would benefit for seeing optionTy and valueOptionTy logic under unified abstractions, since they are coupled together via some features.